base/reduce.jl

```diff
@@ -168,14 +168,17 @@ foldr(op, itr) = mapfoldr(identity, op, itr)
 ## reduce & mapreduce
 
 function mapreduce_impl(f, op, A::AbstractArray, ifirst::Integer, ilast::Integer, blksize::Int=pairwise_blocksize(f, op))
-    if ifirst + blksize > ilast
+    if ifirst == ilast
+        @inbounds a1 = A[ifirst]
```

Comment:
@tkelman I've added a note about `mapreduce_impl()` usage.

P.S. I appreciate your rigorous review as it definitely improves both the PR and my understanding of Julia. ;)